File: src/main/kotlin/com/ederfmatos/reactive/java/One.java
package com.ederfmatos.reactive.java;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public interface One<T> {

    <R> One<R> map(Function<T, R> transformer);

    <R> One<R> flatMap(Function<T, One<R>> transformer);

    <R> One<R> then(Supplier<One<R>> transformer);

    <R> One<R> thenConsume(Supplier<R> transformer);

    One<T> ifEmpty(Supplier<One<T>> supplier);

    One<T> ifEmptyReturn(Supplier<T> supplier);

    One<T> filter(Predicate<T> filter);

    One<T> filterNot(Predicate<T> filter);

    One<T> onNext(Consumer<T> consumer);

    One<T> onError(Consumer<Throwable> consumer);

    One<T> onEmpty(Runnable runnable);

    One<T> onSuccess(Consumer<T> consumer);

    One<T> onComplete(Runnable runnable);

    static <T> One<T> from(T value) {
        return MonoOne.from(value);
    }

    static <T> One<T> empty() {
        return MonoOne.empty();
    }
}

File: src/main/kotlin/com/ederfmatos/reactive/kotlin/One.kt
package com.ederfmatos.reactive.kotlin

interface One<T: Any> {
    fun <R: Any> map(block: (T) -> R): One<R>
    fun <R: Any> flatMap(block: (T) -> One<R>): One<R>
    fun ifEmpty(block: () -> One<T>): One<T>
    fun ifEmptyReturn(block: () -> T): One<T>
    fun filter(block: (T) -> Boolean): One<T>
    fun filterNot(block: (T) -> Boolean): One<T>
    fun onNext(block: (T) -> Unit): One<T>
    fun onError(block: (Throwable) -> Unit): One<T>
    fun onEmpty(block: () -> Unit): One<T>
    fun onSuccess(block: (T) -> Unit): One<T>
    fun onComplete(block: () -> Unit): One<T>

    companion object {
        @JvmStatic
        inline fun <reified T : Any> from(value: T): One<T> {
            return MonoOne.from(value)
        }

        @JvmStatic
        inline fun <reified T : Any> empty(): One<T> {
            return MonoOne.empty()
        }
    }
}